Global ik cost

hongkai-dai opened the pull request:
Emulate the cost in the `inverseKin`, in global IK, add a cost to penalize the deviation of each body pose to the desired body pose.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/5460)
<!-- Reviewable:end -->

hongkai-dai commented:
+@siyuanfeng-tri for feature review please, since you asked the question on the objective function.
+@amcastro-tri for feature review please, since you gave a great feature review on the previous PR and made the code a lot better, it would be great to have your eyes on this PR if you have time.

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-:-KephKfU-L_Zv2Jojf2_:b-pwku4i)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> thanks hongkai! minor comments. i am excited to see what happens if you feed this to snopt. 

---

Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/multibody/global_inverse_kinematics.cc, line 360 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-KeqDMxxUEVP7SZqy21f:-KeqDMxxUEVP7SZqy21g:baqnvdt) ([raw file](https://github.com/robotlocomotion/drake/blob/3f5471a01bd784b3799a514bc0f1e151668352fa/drake/multibody/global_inverse_kinematics.cc#L360)):*
> ```cpp
>       NewContinuousVariables(num_bodies - 1, "p_WBo_error");
>   for (int i = 1; i < num_bodies; ++i) {
>     const auto& T_WB_desired = robot_->CalcFramePoseInWorldFrame(
> ```

BTW, i think the dynamics team likes X better, maybe you have something special for T? 

---

*[drake/multibody/global_inverse_kinematics.cc, line 371 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-KeqQATUnhCqh-3p5G3s:-KeqQATUnhCqh-3p5G3t:b-wj5tim) ([raw file](https://github.com/robotlocomotion/drake/blob/3f5471a01bd784b3799a514bc0f1e151668352fa/drake/multibody/global_inverse_kinematics.cc#L371)):*
> ```cpp
>     // The orientation error is on the angle θ between the body orientation and
>     // the desired orientation, namely 1 - cos(θ).
>     // cos(θ) can be computed as (trace( R_WB_desired * R_WB_[i]ᵀ) - 1) / 2
> ```

cos(theta) = (trace( R_WB_desired * R_WB_[i]ᵀ) - 1) / 2, i think this is from the conversion from angle axis and rotation matrix right? can you add a quick comment here. it took me more than 30sec to realize this.. 

---

*[drake/multibody/global_inverse_kinematics.h, line 156 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-KeqBc3RCWQJ1zekbIQB:-KeqBc3RCWQJ1zekbIQC:board9o) ([raw file](https://github.com/robotlocomotion/drake/blob/3f5471a01bd784b3799a514bc0f1e151668352fa/drake/multibody/global_inverse_kinematics.h#L156)):*
> ```objc
> 
>   /** Penalizes the deviation to the desired posture.
>    * For each body in the kinematic tree, we add the cost
> ```

For body i ? maybe move "Since the body 0 is the world frame, we should not impose a cost
   on the world frame position" part up here would be better? also we are assuming this is a fixed base robot? and there is only one robot in the RBT. 

---

*[drake/multibody/dev/test/global_inverse_kinematics_test.cc, line 168 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-KeqDBTzmXhZF_aGXeP6:-KeqDBU-oopSKFXr-wRi:b-kdgl90) ([raw file](https://github.com/robotlocomotion/drake/blob/3f5471a01bd784b3799a514bc0f1e151668352fa/drake/multibody/dev/test/global_inverse_kinematics_test.cc#L168)):*
> ```cpp
> 
>   Isometry3d ee_desired_pose = rigid_body_tree_->CalcFramePoseInWorldFrame(
>       cache, rigid_body_tree_->get_body(ee_idx_), Isometry3d::Identity());
> ```

BTW, the offset defaults to I, so you could omit it. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-:-KeqQaHJndriYk3LE6Iz:bqgr7oo)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
I will add the snopt IK test in the next PR. 

---

Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/multibody/global_inverse_kinematics.cc, line 360 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-KeqDMxxUEVP7SZqy21f:-KeqtRs8PDDqgRmdcVef:b-yh16l6) ([raw file](https://github.com/robotlocomotion/drake/blob/3f5471a01bd784b3799a514bc0f1e151668352fa/drake/multibody/global_inverse_kinematics.cc#L360)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

BTW, i think the dynamics team likes X better, maybe you have something special for T? 
</blockquote></details>

Done. Good call.

---

*[drake/multibody/global_inverse_kinematics.cc, line 371 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-KeqQATUnhCqh-3p5G3s:-KeqtlHIOSXm4Khbon7r:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3f5471a01bd784b3799a514bc0f1e151668352fa/drake/multibody/global_inverse_kinematics.cc#L371)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

cos(theta) = (trace( R_WB_desired * R_WB_[i]ᵀ) - 1) / 2, i think this is from the conversion from angle axis and rotation matrix right? can you add a quick comment here. it took me more than 30sec to realize this.. 
</blockquote></details>

Done.

---

*[drake/multibody/global_inverse_kinematics.h, line 156 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-KeqBc3RCWQJ1zekbIQB:-KeqttpfiQ40ZLRbdORk:br74vvn) ([raw file](https://github.com/robotlocomotion/drake/blob/3f5471a01bd784b3799a514bc0f1e151668352fa/drake/multibody/global_inverse_kinematics.h#L156)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

For body i ? maybe move "Since the body 0 is the world frame, we should not impose a cost
   on the world frame position" part up here would be better? also we are assuming this is a fixed base robot? and there is only one robot in the RBT. 
</blockquote></details>

Done. I am not assuming it is a fixed base robot. The cost is on any body in the rigid body tree, except for the world. Actually in the kuka example, there is a cost on the fixed base.

I do not have a test for floating base robot yet. Adding a test for a floating base (or even multiple robots) is on my list, but not high priority. I would like to make sure the algorithm works for Kuka arm first, and then think about other examples.

---

*[drake/multibody/dev/test/global_inverse_kinematics_test.cc, line 168 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-KeqDBTzmXhZF_aGXeP6:-Keqv-H7pTL77jKgjRK-:b-oc2vz6) ([raw file](https://github.com/robotlocomotion/drake/blob/3f5471a01bd784b3799a514bc0f1e151668352fa/drake/multibody/dev/test/global_inverse_kinematics_test.cc#L168)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

BTW, the offset defaults to I, so you could omit it. 
</blockquote></details>

Done. Thanks, good to know!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-:-KeqxtWSChhnHiuE1H7F:b-hwt0dt)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Reviewed 1 of 3 files at r2.
Review status: 1 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/multibody/global_inverse_kinematics.h, line 156 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-KeqBc3RCWQJ1zekbIQB:-Kesc5h8X1f3zq2TXrkJ:btlnv56) ([raw file](https://github.com/robotlocomotion/drake/blob/3f5471a01bd784b3799a514bc0f1e151668352fa/drake/multibody/global_inverse_kinematics.h#L156)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Done. I am not assuming it is a fixed base robot. The cost is on any body in the rigid body tree, except for the world. Actually in the kuka example, there is a cost on the fixed base.

I do not have a test for floating base robot yet. Adding a test for a floating base (or even multiple robots) is on my list, but not high priority. I would like to make sure the algorithm works for Kuka arm first, and then think about other examples.
</blockquote></details>

o right, i forgot about the world body. nvm the comment then. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5460)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Reviewed 2 of 3 files at r2.
Review status: 3 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5460)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
+@ggould-tri for platform review please.

---

Review status: 3 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-:-KetFgHWSi0a-qg11Z9v:b-irfx5k)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>  +(status: curate commits before merging)

It took me embarrassingly long to realize the reason you were't putting a cost on body 0 is that it was the world.

---

Reviewed 1 of 4 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/multibody/global_inverse_kinematics.cc, line 325 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-KeuC2hKIn55oyk79VpS:-KeuC2hKIn55oyk79VpT:b-d2w3ww) ([raw file](https://github.com/robotlocomotion/drake/blob/89d01d0b00e1fde137b5e42c5ad1f764d549d6d4/drake/multibody/global_inverse_kinematics.cc#L325)):*
> ```cpp
>     const Eigen::Ref<const Eigen::VectorXd>& body_position_cost,
>     const Eigen::Ref<const Eigen::VectorXd>& body_orientation_cost) {
>   const int num_bodies = robot_->get_num_bodies();
> ```

BTW -- Given that in every use case but one this appears as `num_bodies - 1`, might it make sense to define instead `const int num_costs = robot_->get_num_bodies() - 1  // The number of non-kWorldBodyIndex bodies in the RBT.` and use that instead?

---

*[drake/multibody/global_inverse_kinematics.cc, line 373 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-KeuD9JP-ThhDizbc6sZ:-KeuD9JP-ThhDizbc6s_:bywxwkk) ([raw file](https://github.com/robotlocomotion/drake/blob/89d01d0b00e1fde137b5e42c5ad1f764d549d6d4/drake/multibody/global_inverse_kinematics.cc#L373)):*
> ```cpp
>     // cos(θ) can be computed as (trace( R_WB_desired * R_WB_[i]ᵀ) - 1) / 2
>     // To see how the angle is computed from a rotation matrix, please refer to
>     // http://www.euclideanspace.com/maths/geometry/rotations/conversions/matrixToAngle/
> ```

Thank you for this explanation.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-:-KeuFm0rUOJOCi6gtyPx:bk72hyd)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
@ggould-tri I now require that the input cost vector have the same dimension as the number of bodies (including the world), I feel it is confusing to say that the dimension is different.

---

Review status: 1 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/multibody/global_inverse_kinematics.cc, line 325 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-KeuC2hKIn55oyk79VpS:-KeufA2CWx6Y-EGzLDEm:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/89d01d0b00e1fde137b5e42c5ad1f764d549d6d4/drake/multibody/global_inverse_kinematics.cc#L325)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

BTW -- Given that in every use case but one this appears as `num_bodies - 1`, might it make sense to define instead `const int num_costs = robot_->get_num_bodies() - 1  // The number of non-kWorldBodyIndex bodies in the RBT.` and use that instead?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-:-KeufBJ-6I4f8lzsjhSJ:b-hog98l)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
That makes sense.  I think you have one stale comment as a result, noted below.

---

Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/multibody/global_inverse_kinematics.h, line 170 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-Kf6j_OnDMbC_FlJD0ew:-Kf6j_OocgS6ImwAwfe7:b-4ji2gm) ([raw file](https://github.com/robotlocomotion/drake/blob/1147676c166ef6a48ddbd2bbe55cc1548b3dafc3/drake/multibody/global_inverse_kinematics.h#L170)):*
> ```objc
>    * angle between the orientation of body i'th frame and body i'th frame using
>    * the desired posture.
>    * Notice that since body 0 is the world, the cost on that is omitted.
> ```

BTW -- As I read it, with the code change above this is no longer correct -- this cost is not omitted, it is just ignored.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-:-Kf6jpSudNOpz3dVJyK9:b-92yh6k)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Reviewed 2 of 3 files at r3.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/multibody/global_inverse_kinematics.cc, line 373 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-KeuD9JP-ThhDizbc6sZ:-Kf6tLOZU6JLOy7E3C9s:b-cht3cq) ([raw file](https://github.com/robotlocomotion/drake/blob/89d01d0b00e1fde137b5e42c5ad1f764d549d6d4/drake/multibody/global_inverse_kinematics.cc#L373)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

Thank you for this explanation.
</blockquote></details>

yes! I second that!

---

*[drake/multibody/global_inverse_kinematics.cc, line 359 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-Kf6sXgLUbP3jIVQSyi1:-Kf6sXgLUbP3jIVQSyi2:bghyahk) ([raw file](https://github.com/robotlocomotion/drake/blob/1147676c166ef6a48ddbd2bbe55cc1548b3dafc3/drake/multibody/global_inverse_kinematics.cc#L359)):*
> ```cpp
>   solvers::VectorXDecisionVariable p_WBo_err =
>       NewContinuousVariables(num_bodies - 1, "p_WBo_error");
>   for (int i = 1; i < num_bodies; ++i) {
> ```

BTW, something like `// Skip the world` comment right on the side would be a nice warning comment for most of us that just quickly scan code.

---

*[drake/multibody/global_inverse_kinematics.cc, line 367 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-Kf6vdt0FaNWYagHYOhw:-Kf6vdt0FaNWYagHYOhx:b-z8um3a) ([raw file](https://github.com/robotlocomotion/drake/blob/1147676c166ef6a48ddbd2bbe55cc1548b3dafc3/drake/multibody/global_inverse_kinematics.cc#L367)):*
> ```cpp
>     pos_error_expr << p_WBo_err(i - 1),
>         body_position_cost(i) * (p_WBo_[i] - X_WB_desired.translation());
>     AddLorentzConeConstraint(pos_error_expr);
> ```

So if I understand this correctly then this method does more than just adding a cost, but is also adding this stronger condition on each of the individual poses. Should that be documented in the method's description? 

---

*[drake/multibody/global_inverse_kinematics.h, line 86 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-Kf6nhXq-Mi9OV29M5IE:-Kf6nhXrSWYKXnNL1l2t:b-pmlllc) ([raw file](https://github.com/robotlocomotion/drake/blob/1147676c166ef6a48ddbd2bbe55cc1548b3dafc3/drake/multibody/global_inverse_kinematics.h#L86)):*
> ```objc
>    * where
>    *   - p_WQ is the position of the body point Q measured and expressed in the
>    * world frame `W`.
> ```

woops, it looks like your indentation somehow got ruined here.

---

*[drake/multibody/global_inverse_kinematics.h, line 103 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-Kf6oVVaiKnqc1LuojHv:-Kf6oVVbCwLY0pd2ZEWR:bq2iyjd) ([raw file](https://github.com/robotlocomotion/drake/blob/1147676c166ef6a48ddbd2bbe55cc1548b3dafc3/drake/multibody/global_inverse_kinematics.h#L103)):*
> ```objc
>    * Notice that since the rotation matrix `R_WB` does not lie exactly on the
>    * SO(3), due to the McCormick envelope relaxation, this constraint is subject
>    * to the accumulated error from the root of the kinematics tree.
> ```

BTW, I just realize. Does the error indeed accumulate? I thought these constraints were applied on the "orientation in the world frame matrices" `R_WB` and thus error did not accumulate. Error would indeed accumulate if the constraints were instead applied only to the transformations between bodies, I think.

---

*[drake/multibody/global_inverse_kinematics.h, line 172 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-Kf6qNBEJd3c7kwSxhVp:-Kf6qNBEJd3c7kwSxhVq:b6q3ioi) ([raw file](https://github.com/robotlocomotion/drake/blob/1147676c166ef6a48ddbd2bbe55cc1548b3dafc3/drake/multibody/global_inverse_kinematics.h#L172)):*
> ```objc
>    * Notice that since body 0 is the world, the cost on that is omitted.
>    * @param q_desired  The desired posture.
>    * @param body_position_cost  The cost for each body's position error.
> ```

BTW, I would document here that this has units of `[1/m]` (one over meters). This single fact would make a big difference when selecting these weights. Is there any recommendation on how to select them? for instance, would it be a good idea to select them as one over the desired standard deviation?

---

*[drake/multibody/dev/test/global_inverse_kinematics_test.cc, line 152 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-Kf6rBdLwx1z-MJh5CH_:-Kf6rBdLwx1z-MJh5CHa:b-4sulcc) ([raw file](https://github.com/robotlocomotion/drake/blob/1147676c166ef6a48ddbd2bbe55cc1548b3dafc3/drake/multibody/dev/test/global_inverse_kinematics_test.cc#L152)):*
> ```cpp
> }
> 
> TEST_F(KukaTest, ReachableWithCost) {
> ```

BTW, suggested improvement. This test is actually not testing the possibilities of the new added feature but just that it runs. 
I suggest you put this one single test in its own fixture. Then you can compare the results obtained when "adding cost on the positions" vs. "not adding cost on the positions". Do you get what you expect? you can for instance assert that you get a smaller error on the one position (/orientation) for which you added more cost.

---

*[drake/multibody/dev/test/global_inverse_kinematics_test.cc, line 171 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-Kf6lefm1ZoCfMWYqfFH:-Kf6lefm1ZoCfMWYqfFI:bjel0tx) ([raw file](https://github.com/robotlocomotion/drake/blob/1147676c166ef6a48ddbd2bbe55cc1548b3dafc3/drake/multibody/dev/test/global_inverse_kinematics_test.cc#L171)):*
> ```cpp
>   // Constrain the global IK to reach the exact end effector pose as the
>   // posture q.
>   global_ik_.AddWorldPositionConstraint(ee_idx_, Vector3d::Zero(),
> ```

BTW, probably a good idea/practice to add a comment next to each argument saying what each argument is.

---

*[drake/multibody/dev/test/global_inverse_kinematics_test.cc, line 181 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-Kf6rmGMjRqBPQg0TUQO:-Kf6rmGMjRqBPQg0TUQP:b4kywev) ([raw file](https://github.com/robotlocomotion/drake/blob/1147676c166ef6a48ddbd2bbe55cc1548b3dafc3/drake/multibody/dev/test/global_inverse_kinematics_test.cc#L181)):*
> ```cpp
>   DRAKE_DEMAND(rigid_body_tree_->get_num_bodies() == 12);
>   global_ik_.AddPostureCost(q + q_err, Eigen::VectorXd::Constant(12, 1),
>                             Eigen::VectorXd::Constant(12, 1));
> ```

BTW, unfair here to use "one" for all weights when the first "one" has units of "one over meters" and the other one is dimensionless. It seems to me like the cost on positions is pretty loose with just `1.0 1/m`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5460)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Just a few comments. <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Review status: all files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-:-Kf6w6NQ4ASJzjrlzc11:bl7lixk)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[drake/multibody/global_inverse_kinematics.cc, line 359 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-Kf6sXgLUbP3jIVQSyi1:-Kf7QXKMuEVpCHoJSS4I:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1147676c166ef6a48ddbd2bbe55cc1548b3dafc3/drake/multibody/global_inverse_kinematics.cc#L359)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, something like `// Skip the world` comment right on the side would be a nice warning comment for most of us that just quickly scan code.
</blockquote></details>

Done.

---

*[drake/multibody/global_inverse_kinematics.cc, line 367 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-Kf6vdt0FaNWYagHYOhw:-Kf7QZdV8vHQ-2lM6sj2:b-ksqrjd) ([raw file](https://github.com/robotlocomotion/drake/blob/1147676c166ef6a48ddbd2bbe55cc1548b3dafc3/drake/multibody/global_inverse_kinematics.cc#L367)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

So if I understand this correctly then this method does more than just adding a cost, but is also adding this stronger condition on each of the individual poses. Should that be documented in the method's description? 
</blockquote></details>

This is not a stronger condition. We want to add a cost on the sum of position error  for each body, as `sum_i |p_WB(i) - p_WB_desired(i)|`. We cannot add this sum to the cost function directly, since it is not a linear or quadratic function. So we introduce a slack variable `p_WBo_err(i-1)`, with the Lorentz cone constraint `p_WBo_err(i-1)  >= |p_WBo(i) - p_WBo_err(i-1)|`, and the add the linear cost function `sum_i p_WBo_err(i)`. This is a standard approach to convert non-linear cost function, to linear cost function.

---

*[drake/multibody/global_inverse_kinematics.h, line 86 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-Kf6nhXq-Mi9OV29M5IE:-Kf7RctgfsLx58pVQb1m:b-y5j5r5) ([raw file](https://github.com/robotlocomotion/drake/blob/1147676c166ef6a48ddbd2bbe55cc1548b3dafc3/drake/multibody/global_inverse_kinematics.h#L86)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

woops, it looks like your indentation somehow got ruined here.
</blockquote></details>

I think the doxygen is fine here? The list symbol `-` is compile correctly. This is what doxygen looks like on my screen.
![Screenshot from 2017-03-13 09:17:30.png](https://files.reviewableusercontent.io/images/reviewable/6314000/LsbOMJhQbp1zzuVe_CFZHLx4/Screenshot%20from%202017-03-13%2009%3A17%3A30.png)

---

*[drake/multibody/global_inverse_kinematics.h, line 103 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-Kf6oVVaiKnqc1LuojHv:-Kf7ShVp0jJnbEcYXig0:b-ippqxk) ([raw file](https://github.com/robotlocomotion/drake/blob/1147676c166ef6a48ddbd2bbe55cc1548b3dafc3/drake/multibody/global_inverse_kinematics.h#L103)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, I just realize. Does the error indeed accumulate? I thought these constraints were applied on the "orientation in the world frame matrices" `R_WB` and thus error did not accumulate. Error would indeed accumulate if the constraints were instead applied only to the transformations between bodies, I think.
</blockquote></details>

The way to reconstruct the posture is

1. Get `R_WB` for each body
2. Compute the joint rotation matrix, by `R_PJ = R_WP' * R_WC`. As you can see, the error from the rotation matrix of the parent body `P`, and from the child body `C`, gets accumulated.
3. Compute the joint angle `theta`, that best matches with the joint rotation matrix `R_PJ`.
4. If we do a forward kinematics using this reconstructed posture, since the forward kinematics is computed from the root of the kinematics tree, the error get accumulated again.

---

*[drake/multibody/global_inverse_kinematics.h, line 170 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-Kf6j_OnDMbC_FlJD0ew:-Kf7TgBpt7STGgeou3eC:b-nodu4x) ([raw file](https://github.com/robotlocomotion/drake/blob/1147676c166ef6a48ddbd2bbe55cc1548b3dafc3/drake/multibody/global_inverse_kinematics.h#L170)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

BTW -- As I read it, with the code change above this is no longer correct -- this cost is not omitted, it is just ignored.
</blockquote></details>

Since the world position and orientation error is always 0 (its position is always 0, orientation is always identity, no matter what posture we take), so the cost is always 0. Maybe "omitted" is not the right word, I changed it here.

---

*[drake/multibody/global_inverse_kinematics.h, line 172 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-Kf6qNBEJd3c7kwSxhVp:-Kf7UKJKWyRTJKPsYjSH:b-fr7q04) ([raw file](https://github.com/robotlocomotion/drake/blob/1147676c166ef6a48ddbd2bbe55cc1548b3dafc3/drake/multibody/global_inverse_kinematics.h#L172)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, I would document here that this has units of `[1/m]` (one over meters). This single fact would make a big difference when selecting these weights. Is there any recommendation on how to select them? for instance, would it be a good idea to select them as one over the desired standard deviation?
</blockquote></details>

Thanks, I added the units here.

There is not much guideline to select the cost here. We have chosen a range of cost in the inverse kinematics problem before. The idea is to use different cost value to emphasize on different part of the cost function, so it really depends on what the user want to do with the inverse kinematics problem.

---

*[drake/multibody/dev/test/global_inverse_kinematics_test.cc, line 152 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-Kf6rBdLwx1z-MJh5CH_:-Kf7Xr0wztopwx6HwIb8:betvqfs) ([raw file](https://github.com/robotlocomotion/drake/blob/1147676c166ef6a48ddbd2bbe55cc1548b3dafc3/drake/multibody/dev/test/global_inverse_kinematics_test.cc#L152)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, suggested improvement. This test is actually not testing the possibilities of the new added feature but just that it runs. 
I suggest you put this one single test in its own fixture. Then you can compare the results obtained when "adding cost on the positions" vs. "not adding cost on the positions". Do you get what you expect? you can for instance assert that you get a smaller error on the one position (/orientation) for which you added more cost.
</blockquote></details>

Done. Now test the ik problem with and without the cost, and compare the result.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5460)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Reviewed 4 of 4 files at r4.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5460)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 3 of 4 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/multibody/dev/test/global_inverse_kinematics_test.cc, line 171 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-Kf6lefm1ZoCfMWYqfFH:-Kf7_0KZuli909xmw_qR:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1147676c166ef6a48ddbd2bbe55cc1548b3dafc3/drake/multibody/dev/test/global_inverse_kinematics_test.cc#L171)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, probably a good idea/practice to add a comment next to each argument saying what each argument is.
</blockquote></details>

Done.

---

*[drake/multibody/dev/test/global_inverse_kinematics_test.cc, line 181 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-Kf6rmGMjRqBPQg0TUQO:-Kf7_H2_7Qr_vGu-XTJU:brtgw7x) ([raw file](https://github.com/robotlocomotion/drake/blob/1147676c166ef6a48ddbd2bbe55cc1548b3dafc3/drake/multibody/dev/test/global_inverse_kinematics_test.cc#L181)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, unfair here to use "one" for all weights when the first "one" has units of "one over meters" and the other one is dimensionless. It seems to me like the cost on positions is pretty loose with just `1.0 1/m`.
</blockquote></details>

Hopefully the updated change answers your question? Here actually any cost greater than 0 is sufficient.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5460)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5460)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/multibody/global_inverse_kinematics.cc, line 367 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-Kf6vdt0FaNWYagHYOhw:-Kf7c-FGD07E0-ibRr_i:bub766o) ([raw file](https://github.com/robotlocomotion/drake/blob/1147676c166ef6a48ddbd2bbe55cc1548b3dafc3/drake/multibody/global_inverse_kinematics.cc#L367)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

This is not a stronger condition. We want to add a cost on the sum of position error  for each body, as `sum_i |p_WB(i) - p_WB_desired(i)|`. We cannot add this sum to the cost function directly, since it is not a linear or quadratic function. So we introduce a slack variable `p_WBo_err(i-1)`, with the Lorentz cone constraint `p_WBo_err(i-1)  >= |p_WBo(i) - p_WBo_err(i-1)|`, and the add the linear cost function `sum_i p_WBo_err(i)`. This is a standard approach to convert non-linear cost function, to linear cost function.
</blockquote></details>

Ah!, I see, thanks

---

*[drake/multibody/global_inverse_kinematics.h, line 86 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-Kf6nhXq-Mi9OV29M5IE:-Kf7c9gss0DjXPB8YuTx:b-qa2o32) ([raw file](https://github.com/robotlocomotion/drake/blob/1147676c166ef6a48ddbd2bbe55cc1548b3dafc3/drake/multibody/global_inverse_kinematics.h#L86)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I think the doxygen is fine here? The list symbol `-` is compile correctly. This is what doxygen looks like on my screen.
![Screenshot from 2017-03-13 09:17:30.png](https://files.reviewableusercontent.io/images/reviewable/6314000/LsbOMJhQbp1zzuVe_CFZHLx4/Screenshot%20from%202017-03-13%2009%3A17%3A30.png)

</blockquote></details>

oh, I didnt mean the Doxygen rendered documentation but the text in the code. I just wonder if the space changes were intentional. It looked better before. Otherwise you might want to indent them with extra spaces. 

---

*[drake/multibody/global_inverse_kinematics.h, line 103 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-Kf6oVVaiKnqc1LuojHv:-Kf7cmOjY5qAn8-nPRSc:b-abyuvv) ([raw file](https://github.com/robotlocomotion/drake/blob/1147676c166ef6a48ddbd2bbe55cc1548b3dafc3/drake/multibody/global_inverse_kinematics.h#L103)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

The way to reconstruct the posture is

1. Get `R_WB` for each body
2. Compute the joint rotation matrix, by `R_PJ = R_WP' * R_WC`. As you can see, the error from the rotation matrix of the parent body `P`, and from the child body `C`, gets accumulated.
3. Compute the joint angle `theta`, that best matches with the joint rotation matrix `R_PJ`.
4. If we do a forward kinematics using this reconstructed posture, since the forward kinematics is computed from the root of the kinematics tree, the error get accumulated again.
</blockquote></details>

I can see how the error accumulates when reconstructing the vector of generalized coordinates (not the pose). I thought the poses themselves were a direct solution of the problem. Actually I just realize that `ReconstructPostureSolution()` should probably instead be called something like `ReconstructGeneralizedPositionsVector()`. The error does accumulate there as you point out. 
If all you want to do is to reconstruct the actual spatial poses, then by just returning the projection of `GetSolution(R_WB_[body_idx]);` into valid rotation matrices would suffice. There would be the error involved in this projection, but no accumulation.
If a user now does need the vector of generalized positions then yes, you have no other option than accumulating these errors. Just an idea.

---

*[drake/multibody/global_inverse_kinematics.h, line 167 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-Kf7eDnm2H_NRPGN6Lqv:-Kf7eDnm2H_NRPGN6Lqw:b8f3bxi) ([raw file](https://github.com/robotlocomotion/drake/blob/bf5502543b49a88388899c41f01c976617b55cab/drake/multibody/global_inverse_kinematics.h#L167)):*
> ```objc
>    *                     `W`, computed from the desired posture `q_desired`.
>    *
>    * body_orientation_error(i) is computed as (1 - cos(θ)), where θ is the
> ```

BTW, I had another thought on this that I wanted to share with you. Notice that in general (hopefully) this `theta` will be a small angle. Therefore we could (just to get a feeling of what that does for us) approximate `(1 - cos(θ)) ≈ θ²/2 + O(θ⁴)`.
What seems to tell us is that `body_orientation_cost` actually scales as a cost on the square of the angle difference rather than on just the difference. This is in contrast to the position error that does effectively act on the distance error (rather than the square). Anyways, thinking about units got me here. 

---

*[drake/multibody/dev/test/global_inverse_kinematics_test.cc, line 199 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-Kf7gJkdGnK_RwNW9Shg:-Kf7gJkdGnK_RwNW9Shh:b-7lb7o6) ([raw file](https://github.com/robotlocomotion/drake/blob/bf5502543b49a88388899c41f01c976617b55cab/drake/multibody/dev/test/global_inverse_kinematics_test.cc#L199)):*
> ```cpp
>     // being equal to q.
>     global_ik_.AddPostureCost(q, Eigen::VectorXd::Constant(12, 1),
>                             Eigen::VectorXd::Constant(12, 1));
> ```

BTW, odd indentation.

---

*[drake/multibody/dev/test/global_inverse_kinematics_test.cc, line 218 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-Kf7h27YzRLX3DDcch2U:-Kf7h27YzRLX3DDcch2V:bgmy920) ([raw file](https://github.com/robotlocomotion/drake/blob/bf5502543b49a88388899c41f01c976617b55cab/drake/multibody/dev/test/global_inverse_kinematics_test.cc#L218)):*
> ```cpp
>         CompareMatrices(q_w_cost, q, 1E-2, MatrixCompareType::absolute));
>     // The posture from IK with cost should be different from that without cost.
>     EXPECT_GE((q_w_cost - q_no_cost).norm(), 1E-3);
> ```

This is great! but was I was trying to say is that it'd be nice to have right below a test like:
`EXPECT_LT((q_w_cost - q).norm(), (q_no_cost - q).norm());`
therefore demonstrating the cost has the desired effect.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5460)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 1 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/multibody/global_inverse_kinematics.h, line 86 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-Kf6nhXq-Mi9OV29M5IE:-Kf7oJBURcTWSQVVU5LA:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1147676c166ef6a48ddbd2bbe55cc1548b3dafc3/drake/multibody/global_inverse_kinematics.h#L86)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

oh, I didnt mean the Doxygen rendered documentation but the text in the code. I just wonder if the space changes were intentional. It looked better before. Otherwise you might want to indent them with extra spaces. 
</blockquote></details>

Done.

---

*[drake/multibody/global_inverse_kinematics.h, line 103 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-Kf6oVVaiKnqc1LuojHv:-Kf7ofWTlIYVprRZMxe6:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1147676c166ef6a48ddbd2bbe55cc1548b3dafc3/drake/multibody/global_inverse_kinematics.h#L103)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I can see how the error accumulates when reconstructing the vector of generalized coordinates (not the pose). I thought the poses themselves were a direct solution of the problem. Actually I just realize that `ReconstructPostureSolution()` should probably instead be called something like `ReconstructGeneralizedPositionsVector()`. The error does accumulate there as you point out. 
If all you want to do is to reconstruct the actual spatial poses, then by just returning the projection of `GetSolution(R_WB_[body_idx]);` into valid rotation matrices would suffice. There would be the error involved in this projection, but no accumulation.
If a user now does need the vector of generalized positions then yes, you have no other option than accumulating these errors. Just an idea.
</blockquote></details>

Done.

---

*[drake/multibody/global_inverse_kinematics.h, line 167 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-Kf7eDnm2H_NRPGN6Lqv:-Kf7nRQA3RagmUO9asbS:b-44uum2) ([raw file](https://github.com/robotlocomotion/drake/blob/bf5502543b49a88388899c41f01c976617b55cab/drake/multibody/global_inverse_kinematics.h#L167)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, I had another thought on this that I wanted to share with you. Notice that in general (hopefully) this `theta` will be a small angle. Therefore we could (just to get a feeling of what that does for us) approximate `(1 - cos(θ)) ≈ θ²/2 + O(θ⁴)`.
What seems to tell us is that `body_orientation_cost` actually scales as a cost on the square of the angle difference rather than on just the difference. This is in contrast to the position error that does effectively act on the distance error (rather than the square). Anyways, thinking about units got me here. 
</blockquote></details>

That is a good analysis, thanks!

---

*[drake/multibody/dev/test/global_inverse_kinematics_test.cc, line 199 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-Kf7gJkdGnK_RwNW9Shg:-Kf7nsL9Q-_LbMujt8c6:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/bf5502543b49a88388899c41f01c976617b55cab/drake/multibody/dev/test/global_inverse_kinematics_test.cc#L199)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, odd indentation.
</blockquote></details>

Done.

---

*[drake/multibody/dev/test/global_inverse_kinematics_test.cc, line 218 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-Kf7h27YzRLX3DDcch2U:-Kf7mLjt2uiL6asdzwRV:bk8q7zs) ([raw file](https://github.com/robotlocomotion/drake/blob/bf5502543b49a88388899c41f01c976617b55cab/drake/multibody/dev/test/global_inverse_kinematics_test.cc#L218)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

This is great! but was I was trying to say is that it'd be nice to have right below a test like:
`EXPECT_LT((q_w_cost - q).norm(), (q_no_cost - q).norm());`
therefore demonstrating the cost has the desired effect.
</blockquote></details>

I think the current test is stronger? Basically I test
`EXPECT_EQ(q_w_cost, q)`
and 
`EXPECT_NE(q_no_cost, q)`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5460)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Reviewed 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5460)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/multibody/dev/test/global_inverse_kinematics_test.cc, line 218 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-Kf7h27YzRLX3DDcch2U:-Kf887EnLdnQyeb-JxeC:b-gv6huo) ([raw file](https://github.com/robotlocomotion/drake/blob/bf5502543b49a88388899c41f01c976617b55cab/drake/multibody/dev/test/global_inverse_kinematics_test.cc#L218)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I think the current test is stronger? Basically I test
`EXPECT_EQ(q_w_cost, q)`
and 
`EXPECT_NE(q_no_cost, q)`.
</blockquote></details>

Not really. The second test really doesn't tell you anything at all. Also the norms used by `CompareMatrices()` and `norm()` are different.
But it's not about those tests. Leave them as they are if you want. All I'm saying is that I think you'd expect the solution to the problem with the additional cost to be closer to the desired state. Then I think it would be fair to check `EXPECT_LT((q_w_cost - q).norm(), (q_no_cost - q).norm());`. Not only fair, but probably the first thing you'd like to check to verify the correct behavior of the implementation. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5460)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5460)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/multibody/dev/test/global_inverse_kinematics_test.cc, line 218 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-Kf7h27YzRLX3DDcch2U:-Kf8O8hVlVfSn70jC8nQ:b-qr54ic) ([raw file](https://github.com/robotlocomotion/drake/blob/bf5502543b49a88388899c41f01c976617b55cab/drake/multibody/dev/test/global_inverse_kinematics_test.cc#L218)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Not really. The second test really doesn't tell you anything at all. Also the norms used by `CompareMatrices()` and `norm()` are different.
But it's not about those tests. Leave them as they are if you want. All I'm saying is that I think you'd expect the solution to the problem with the additional cost to be closer to the desired state. Then I think it would be fair to check `EXPECT_LT((q_w_cost - q).norm(), (q_no_cost - q).norm());`. Not only fair, but probably the first thing you'd like to check to verify the correct behavior of the implementation. 
</blockquote></details>

Ah! it took me so much brain power to get this thing straight! thank you for the VC explanation @hongkai-dai 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5460)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
-(status: curate commits before merging) 

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5460#-:-Kf8OQLt5dCUJWresYf6:b-l70wzf)*
<!-- Sent from Reviewable.io -->
